core/plug-in/codec2/codec2.c

```diff
@@ -0,0 +1,299 @@
+
+  struct codec2_encoder* c2enc = (struct codec2_encoder*)h_inst;
+
+  codec2_destroy(c2enc->codec2);
+  free(c2enc);
+}
+
+
+
+/**
+ * Function encodes buffer using Codec2 library.
+ *
+ * Arguments:
+ * - out_buf: An output buffer where will be written encoded data.
+ * - in_buf: A raw data, which will be encoded using Codec2's encode function.
+ * - size: Size of in_buf.
+ * - channels: A channel, which must be 1 as we use one channel.
+ * - rate: It must be 8000.
+ * - h_codec: An address of codec2 structure created by sems_codec2_create function.
+ *
+ * Returns a length of encoded data in bytes.
+ */
+int pcm16_2_codec2(unsigned char* out_buf, unsigned char* in_buf, unsigned int size,
+                   unsigned int channels, unsigned int rate, long h_codec) {
+
+  // Codec2 encode;
+
+  if (!h_codec) {
+    ERROR("Codec2 codec was not initialized.\n");
+    return -1;
+  }
+
+  if ((channels != 1) || (rate != 8000)) {
+    ERROR("Unsupported input format for Codec2 encoder.\n");
+    return -1;
+  }
+
+  struct codec2_encoder* c2enc = (struct codec2_encoder*)h_codec;
+  struct CODEC2* codec2 = c2enc->codec2;
+
+  const int in_sample_count = size / 2;
+  if (in_sample_count < c2enc->samples_per_frame) {
+    ERROR("Size of input buffer's frame is less than size of frame in codec2.\n");
+    return -1;
+  }
+
+  div_t blocks = div(in_sample_count, c2enc->samples_per_frame);
+  if (blocks.rem) {
+    ERROR("pcm16_2_codec2: not integral number of blocks %d.%d\n", blocks.quot, blocks.rem);
+    return -1;
+  }
+
+  // For each chunk of c2enc->samples_per_frame bytes, encode a single frame.
+  int out_buffer_offset = 0;
+  int in_buffer_offset = 0;
+  const int in_buffer_offset_next = c2enc->samples_per_frame * 2;
+
+  while (blocks.quot--) {
+    codec2_encode(codec2, out_buf + out_buffer_offset, in_buf + in_buffer_offset);
+    out_buffer_offset += c2enc->nbyte;
+    in_buffer_offset += in_buffer_offset_next;
+  }
+
+  return out_buffer_offset;
+}
+
+
+
+/**
+ * Function decodes received data using Codec2 library.
+ *
+ * Arguments:
+ * - out_buf: An output buffer where will be written decoded data.
+ * - in_buf: An encoded data, which will be decoded using Codec2's decode function.
+ * - size: Size of in_buf.
+ * - channels: A channel, which must be 1 as we use one channel.
+ * - rate: It must be 8000.
+ * - h_codec: An address of codec2 structure created by sems_codec2_create function.
+ */
+int codec2_2_pcm16(unsigned char* out_buf, unsigned char* in_buf, unsigned int size,
+                   unsigned int channels, unsigned int rate, long h_codec) {
+
+  // Codec2 decode;
+
+  if (!h_codec) {
+    ERROR("Codec2 codec was not initialized.\n");
+    return -1;
+  }
+
+  struct codec2_encoder* c2enc = (struct codec2_encoder*)h_codec;
+  struct CODEC2* codec2 = c2enc->codec2;
+
+  div_t blocks = div(size, c2enc->nbyte);
+  if (blocks.rem) {
+    ERROR("pcm16_2_codec2: not integral number of blocks %d.%d\n", blocks.quot, blocks.rem);
```

Comment:
We're in codec2_2_pcm16 here, not in pcm16_2_code2. It would be nice to update the formatting string.